blackfynn/models/workspace.py

```
@@ -139,6 +143,46 @@ def get_all_named_queries(self):
         """ Retrieve all named queries in a workspace """
         return self._api.analytics.get_all_named_queries(self)
 
+    def execute_query(self, query_str):
+        """ Execute a query against a snapshot in the workspace
+
+        Args:
+            query_str (str): Query to execute
+        Returns:
+            str
+        """
+        return self._api.analytics.execute_query(self, query_str)
+
+    def get_query_execution_results(self, execution_id):
+        """ Get results from an executed query
+         Args:
+            execution_id (str): Query execution ID
+        Returns:
+            QueryResults
+        """
+        return self._api.analytics.get_query_execution_results(
+            self, execution_id)
+
+    def get_query_execution_status(self, execution_id):
+        """ Get execution status for a query
+        Args:
+            execution_id (str): Query execution ID
+        Returns:
+            str
+        """
+        return self._api.analytics.get_query_execution_status(self,
+                                                              execution_id)
+
+    def get_query_results_presigned_url(self, execution_id):
+        """ Get download link for the query results
+        Args:
+            execution_id (str): Query execution ID
+        Returns:
+            str
+        """
+        return self._api.analytics.get_query_results_presigned_url(
+            self, execution_id)
+
```

I want to simplify this API. Could something like the following work?

- `self._api.analytics.execute_query` returns the `QueryResults` object or a wrapper that just tracks an `execution_id`
- The `QueryResults` then has a `status` property/method (that calls `get_query_execution_status`) and the `download` and `as_dataframe` methods. These should raise an exception if the results are not ready. (I'm not sure if this requires you to get the `status` first, or if the results endpoint gives back that information.)

This would make it so that `NamedQuery.execute` and `Workspace.execute_query` return the same objects and make the user flow look something like this:

```
results = named_query.execute()
results = workspace.execute_query("SELECT * ...")  

results.status                # Dynamically retrieved
results.download()        # exception if not available
results.as_dataframe()  # exception if not available
```

I think this would be a lot cleaner than having users know about `execution_id`.

---

Yep, that is a cleaner API from the user perspective. I was trying to keep the extra work that the python client is doing to a minimum so that it is more of a basic client to interact with the API as it is defined, which requires working with the `execution_id`. Requiring the user to keep track of the execution id is also how the API for the Athena client works. I guess this is more of a philosophical difference as to what a client should be doing. I prefer the minimalist approach so that the language-specific clients are a simple wrapper over the actual API. That said, offering a more intuitive interface is more consistent how the rest of the python client works.

Some day, I'd like to take the AWS approach and have two ways to interact with our APIs: one low-level client interface that is this really thin wrapper around the API and then a higher-level interface for interacting with resources on the platform via language-specific objects. It would make it easier to auto-generate the lower-level APIs based on swagger or some other IDL. I know this separation exists in the python client today, we just don't give users explicit access to the lower-level api.